pyk/src/tests/integration/proof/test_custom_step.py

```diff
@@ -0,0 +1,232 @@
+
+    def abstract_node(self, c: CTerm) -> CTerm:
+        return c
+
+    def same_loop(self, c1: CTerm, c2: CTerm) -> bool:
+        return False
+
+    def custom_step(self, c: CTerm) -> KCFGExtendResult | None:
+        return None
+
+
+class CustomStepSemanticsWithStep(CustomStepSemanticsWithoutStep):
+    def custom_step(self, c: CTerm) -> KCFGExtendResult | None:
+        k_cell = c.cell('K_CELL')
+        if (
+            type(k_cell) is KSequence
+            and type(k_cell[0]) is KApply
+            and k_cell[0].label.name == 'c_CUSTOM-STEP-SYNTAX_Step'
+        ):
+            new_cterm = CTerm.from_kast(set_cell(c.kast, 'K_CELL', KSequence(KApply('d_CUSTOM-STEP-SYNTAX_Step'))))
+            return Step(new_cterm, 1, (), ['CUSTOM-STEP.c.d'], cut=True)
+        return None
+
+
+def cterm_template(label: str) -> CTerm:
+    return CTerm(
+        KApply(
+            KLabel('<generatedTop>'),
+            [
+                KApply(
+                    KLabel('<k>'),
+                    [
+                        KSequence(KApply(KLabel(label))),
+                    ],
+                ),
+                KApply(KLabel('<generatedCounter>'), [KVariable('GENERATEDCOUNTER_CELL', KSort(name='Int'))]),
+            ],
+        ),
+    )
+
+
+CUSTOM_STEP_TEST_DATA_APPLY: Iterable[tuple[str, CTerm, KCFGExtendResult | None]] = (
+    (
+        'None',
+        cterm_template('a_CUSTOM-STEP-SYNTAX_Step'),
+        None,
+    ),
+    (
+        'Step',
+        cterm_template('c_CUSTOM-STEP-SYNTAX_Step'),
+        Step(
+            cterm_template('d_CUSTOM-STEP-SYNTAX_Step'),
+            1,
+            (),
+            ['CUSTOM-STEP.c.d'],
+            cut=True,
+        ),
+    ),
+)
+
+
+class TestCustomStep(CTermSymbolicTest, KProveTest):
+    KOMPILE_DEFINITION = """
+        module CUSTOM-STEP-SYNTAX
+
+            syntax KItem ::= Step
+            syntax Step ::= "a" | "b" | "c" | "d"
+        endmodule
+
+        module CUSTOM-STEP
+            imports CUSTOM-STEP-SYNTAX
+            imports K-EQUAL
+            configuration <k> $PGM:KItem </k>
+
+            rule [a.b]: <k> a => b ... </k>
+            rule [b.c]: <k> b => c ... </k>
+            rule [c.d]: <k> c => d ... </k>
+        endmodule
+    """
+    KOMPILE_MAIN_MODULE = 'CUSTOM-STEP'
+
+    @pytest.fixture
+    def create_kcfg_explore(
+        self,
+        cterm_symbolic: CTermSymbolic,
+        bug_report: BugReport | None,
+    ) -> Callable[[KCFGSemantics], KCFGExplore]:
+        def _create_kcfg_explore(kcfg_semantics: KCFGSemantics) -> KCFGExplore:
+            return KCFGExplore(cterm_symbolic, kcfg_semantics=kcfg_semantics)
+
+        return _create_kcfg_explore
+
+    @pytest.mark.parametrize(
+        'test_id,cterm,expected',
+        CUSTOM_STEP_TEST_DATA_APPLY,
+        ids=[test_id for test_id, *_ in CUSTOM_STEP_TEST_DATA_APPLY],
+    )
+    def test_custom_step_exec(self, test_id: str, cterm: CTerm, expected: KCFGExtendResult | None) -> None:
```

Comment:
This test is a bit superfluous this way I think. What would make sense is to test `custom_step` in integration with `KCFGExplore`, i.e. extend a `KCFG` a single step and check the result.

Reply:
@tothtamas28  I'd like to merge this as it is for now to unblock https://github.com/runtimeverification/evm-semantics/pull/2441. Also, I'd say that the integration with `KCFGExplore` is tested using the `_run_prover` method below. I can open a follow-up PR to change this test if needed.